### PR TITLE
[uss_qualifier/common_dict_eval] Move speed to generic function

### PR DESCRIPTION
**Pull request by the-glu**
With tests :)

Following @BenjaminPelletier remarks in #940, this moves speed checks to the `_generic_evaluator`, improving tests and edge case handeling.

### CODE REVIEW

File: monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator_test.py [hunk 1]
```diff
@@ -300,17 +300,6 @@ def step_under_test(self: UnitTestScenario):
...
 
-def test_speed():
-    _assert_speed(1, 1, True)  # Ok
-    _assert_speed(20.75, 20.75, True)  # Ok
-    _assert_speed(400, 400, False)  # Fail, above MaxSpeed
-    _assert_speed(23.3, 23.3, True)  # Ok
-    _assert_speed(23.13, 23.25, True)  # Ok
-    _assert_speed(23.12, 23.0, True)  # Ok
-    _assert_speed(23.13, 23.0, False)  # Ok
-    _assert_speed(23.13, 23.5, False)  # Ok
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator_test.py:311]: Clean up `_assert_speed`?

File: monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator_test.py [hunk 2]
```diff
@@ -1210,3 +1202,76 @@ def dp_field_setter(flight: Any, value: T) -> Any:
...
+        dp_field_setter,
+    )
+
+    _assert_generic_evaluator_correct_field_is_used(
+        *base_args,
+        valid_value=42,
+        valid_value_2=3.14,
+    )
+
+    # Value should be betewwen -254.25, 254.25 or the special 255 value
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator_test.py:1235]:
```suggestion
    # Value should be between -254.25, 254.25 or the special 255 value
```

File: monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py [hunk 3]
```diff
@@ -426,52 +417,51 @@ def _evaluate_operator_id(
...
             )
 
+        def value_validator(val: float) -> float:
+            if val == constants.SpecialSpeed:
+                return val
+
+            if val < -constants.MaxSpeed:
+                raise ValueError(f"Speed is less than -{constants.MaxSpeed}")
+            if val > constants.MaxSpeed:
+                raise ValueError(f"Speed is greather than {constants.MaxSpeed}")
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py:443]:
```suggestion
                raise ValueError(f"Speed is greater than {constants.MaxSpeed}")
```

File: monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py [hunk 4]
```diff
@@ -426,52 +417,51 @@ def _evaluate_operator_id(
...
-                    "Observed speed is consistent with injected one", participants
-                ) as check:
-                    # Speed seems rounded to nearest 0.25 m/s -> x.0, x.25, x.5, x.75, (x+1).0
-                    if abs(speed_obs - speed_inj) > 0.125:
-                        check.record_failed(
-                            "Observed speed different from injected speed",
-                            details=f"Injected speed was {speed_inj} - observed speed is {speed_obs}",
-                        )
-        else:
+        if self._rid_version != RIDVersion.f3411_22a:
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py:429]: Speed exists in v19 doesn't it?

File: monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py [hunk 5]
```diff
@@ -426,52 +417,51 @@ def _evaluate_operator_id(
...
             self._test_scenario.record_note(
                 key="skip_reason",
                 message=f"Unsupported version {self._rid_version}: skipping speed evaluation",
             )
 
+        def value_validator(val: float) -> float:
+            if val == constants.SpecialSpeed:
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py:437]: Compare float with `math.isclose`

File: monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py [hunk 6]
```diff
@@ -426,51 +417,51 @@ def _evaluate_operator_id(
...
+            if val > constants.MaxSpeed:
+                raise ValueError(f"Speed is greather than {constants.MaxSpeed}")
+            return val
+
+        def value_comparator(v1: Optional[float], v2: Optional[float]) -> bool:
+
+            if v1 is None or v2 is None:
+                return False
+
+            return abs(v1 - v2) < constants.MinSpeedResolution
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py:451]: If my understanding is correct: the SP or DP may round the value up to the minimum speed resolution. In that case shouldn't the absolute difference be maximum half of the minimum speed resolution?
(also why before the comparison was done with *0.125*)

File: monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py [hunk 7]
```diff
@@ -426,51 +417,51 @@ def _evaluate_operator_id(
...
+
+        self._generic_evaluator(
+            "speed",
+            "raw.current_state.speed",
+            "current_state.speed",
+            "Speed",
+            value_validator,
+            None,
+            True,
```
Comment (mickmis) [inline monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py:460]: In the end, it is not actually required to be injected, is it?

Comment (the-glu) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py:460]: They are for a "successful injection" so that correct, if I'm not wrong. It's the same as others fields like vertical_speed

File: monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py (hunk 4 above)
Comment (the-glu) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py:429]: Yes I was not sure why the old function has this check, I was already testing locally and it does seems to works, so it's now in v19 :)

File: monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py (hunk 6 above)
Comment (the-glu) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py:451]: (After a side discussion, 0.125 doesn't works and is probably wrong ^^')

File: monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py (hunk 7 above)
Comment (mickmis) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py:453]: right

**Review by mickmis — APPROVED** [review]
LGTM modulo one issue

File: monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py [hunk 8]
```diff
@@ -426,51 +417,44 @@ def _evaluate_operator_id(
...
-        else:
-            self._test_scenario.record_note(
-                key="skip_reason",
-                message=f"Unsupported version {self._rid_version}: skipping speed evaluation",
-            )
+        def value_validator(val: float) -> float:
+            if math.isclose(val, constants.SpecialSpeed):
+                return val
+
+            if val < -constants.MaxSpeed:
```
[the remark this answers is not in the document]
Comment (the-glu) [reply monitoring/uss_qualifier/scenarios/astm/netrid/common_dictionary_evaluator.py]: True ^^'
